ansible_runner/interface.py

```diff
@@ -91,7 +93,15 @@ def init_runner(**kwargs):
     event_callback_handler = kwargs.pop('event_handler', None)
     status_callback_handler = kwargs.pop('status_handler', None)
     artifacts_handler = kwargs.pop('artifacts_handler', None)
-    cancel_callback = kwargs.pop('cancel_callback', None)
+    if kwargs.get('cancel_callback'):
+        cancel_callback = kwargs.pop('cancel_callback')
+    # Only the main thread is allowed to set a new signal handler
+    elif threading.current_thread() is threading.main_thread():
+        signal_handler = SignalHandler()
+        cancel_callback = signal_handler.called
+    else:
+        logger.warning("Unable to set cancel_callback for signal handling; not in main thread")
+        cancel_callback = None
```

Comment:
I'm wrestling with this one a bit. I tried this up so that we had a flag indicating whether or not the warning has been issued already. In the scenario you outlined, though, the warning is being generated anytime we're in a new thread. That seems to suggest that when we're in a thread, we need to set a flag so that future threads know not to send a warning. It seems like this would require sending a message via a queue, I'm not sure. Again, I'm just not exactly sure how we would do this. Do you have thoughts on how to make this work? How strongly do you feel about having this warning? It seems like it might be fine to leave it out. Just adding the signal handling behavior by itself is a big improvement, even if we can't always set one (and even if we don't notify the user that they're missing out).

Reply:
✅